tests/unit/s2n_cert_validation_callback_test.c

```diff
@@ -0,0 +1,414 @@
+                EXPECT_FAILURE_WITH_ERRNO(s2n_cert_validation_accept(&info), S2N_ERR_INVALID_STATE);
+            }
+
+            /* State was updated from the successful call */
+            EXPECT_EQUAL(info.finished, true);
+            EXPECT_EQUAL(info.accepted, false);
+        }
+    }
+
+    /* Test s2n_cert_validation_callback */
+    {
+        /* clang-format off */
+        struct {
+            struct cert_validation_data data;
+            s2n_error expected_error;
+        } test_cases[] = {
+            /* No error when accept is called from the callback */
+            {
+                .data = { .call_accept_or_reject = true, .accept = true, .return_success = true },
+                .expected_error = S2N_ERR_OK
+            },
+
+            /* Error if reject was called from the callback */
+            {
+                .data = { .call_accept_or_reject = true, .accept = false, .return_success = true },
+                .expected_error = S2N_ERR_CERT_VALIDATION_CALLBACK_FAILED
+            },
+
+            /* Error if the callback doesn't return successfully */
+            {
+                .data = { .call_accept_or_reject = true, .accept = true, .return_success = false },
+                .expected_error = S2N_ERR_CANCELLED
+            },
+            {
+                .data = { .call_accept_or_reject = true, .accept = false, .return_success = false },
+                .expected_error = S2N_ERR_CANCELLED
+            },
+            {
+                .data = { .call_accept_or_reject = false, .return_success = false },
+                .expected_error = S2N_ERR_CANCELLED
+            },
+
+            /* Error if accept or reject wasn't called from the callback */
+            {
+                .data = { .call_accept_or_reject = false, .return_success = true },
+                .expected_error = S2N_ERR_INVALID_STATE
+            },
+        };
+        /* clang-format on */
+
+        /* s2n_x509_validator test */
+        for (int i = 0; i < s2n_array_len(test_cases); i++) {
+            DEFER_CLEANUP(struct s2n_x509_trust_store trust_store = { 0 }, s2n_x509_trust_store_wipe);
+            s2n_x509_trust_store_init_empty(&trust_store);
+
+            char cert_chain[S2N_MAX_TEST_PEM_SIZE] = { 0 };
+            EXPECT_SUCCESS(s2n_read_test_pem(S2N_DEFAULT_TEST_CERT_CHAIN, cert_chain, S2N_MAX_TEST_PEM_SIZE));
+            EXPECT_SUCCESS(s2n_x509_trust_store_add_pem(&trust_store, cert_chain));
+
+            DEFER_CLEANUP(struct s2n_x509_validator validator, s2n_x509_validator_wipe);
+            EXPECT_SUCCESS(s2n_x509_validator_init(&validator, &trust_store, 0));
+
+            DEFER_CLEANUP(struct s2n_config *config = s2n_config_new(), s2n_config_ptr_free);
+            EXPECT_NOT_NULL(config);
+            EXPECT_SUCCESS(s2n_config_set_cipher_preferences(config, "default"));
+
+            DEFER_CLEANUP(struct cert_validation_data *data = &test_cases[i].data, reset_cert_validation_data);
```

Comment:
This seems unsafe. Tests shouldn't affect each other. Why not make the cert_validation_data in the test case const, and do `struct cert_validation_data data = test_cases[i].data`?

Alternatively, have a different structure include the const structure from the test_case + a mutable count.